src/transports/utils/add_query_string.ts

```diff
@@ -0,0 +1,71 @@
+/**
+ * Add to an URL a query string corresponding to `supplementaryQueryStringData`,
+ * being tuples where the first element is a query string's property name and the
+ * second element is its value (or null if there's no associated value.
+ *
+ * If the given URL already has a query string, the new query string elements
+ * will just be appended.
+ * @param {string} baseUrl
+ * @param {Array.<Array.<string|null>>} supplementaryQueryStringData
+ * @returns {string}
+ */
+export default function addQueryString(
+  baseUrl: string,
+  supplementaryQueryStringData: Array<[string, string | null]>,
+): string {
+  if (supplementaryQueryStringData.length === 0) {
+    return baseUrl;
+  }
+  let urlPrefix: string;
+  let urlSuffix: string | null;
+  let queryStringStartingChar: "?" | "&" | "";
+  const indexOfQueryString = baseUrl.indexOf("?");
+  if (indexOfQueryString === -1) {
+    queryStringStartingChar = "?";
+    const indexOfFragment = baseUrl.indexOf("#");
+    if (indexOfFragment >= 0) {
+      urlPrefix = baseUrl.substring(0, indexOfFragment);
+      urlSuffix = baseUrl.substring(indexOfFragment);
+    } else {
+      urlPrefix = baseUrl;
+      urlSuffix = null;
+    }
+  } else {
+    let indexOfFragment = baseUrl.substring(indexOfQueryString).indexOf("#");
+    if (indexOfFragment >= 0) {
+      indexOfFragment += indexOfQueryString;
+      if (indexOfQueryString + 1 === indexOfFragment) {
+        queryStringStartingChar = "";
+      } else {
+        queryStringStartingChar = "&";
+      }
+      urlPrefix = baseUrl.substring(0, indexOfFragment);
+      urlSuffix = baseUrl.substring(indexOfFragment);
+    } else {
+      if (indexOfQueryString === baseUrl.length - 1) {
+        queryStringStartingChar = "";
+      } else {
+        queryStringStartingChar = "&";
+      }
+      urlPrefix = baseUrl;
+      urlSuffix = null;
+    }
```

Comment:
Parsing the fragment can be done outside of the if/else `indexOfQueryString === -1`.

It would have less branching like that:

```js
const indexOfFragment = baseUrl.indexOf("#");
const fragmentSring = baseUrl.substring(indexOfFragment, baseUrl.length);
baseUrl = baseUrl.substring(0, indexOfFragment);
...
return url + fragmentString

```

Reply:
I'm not sure I follow how this example would be integrated but I will re-check.

Though that example at line 2 wouldn't work if no fragment is in that URL

Reply:
I need to rewrite the fragment checking logic anyway so it comes before the querystring parsing  I guess

Reply:
I updated the code. Is it better now?

Reply:
Yes perfect it's better